lib/doorkeeper/oauth/authorization_code_request.rb

```diff
@@ -6,18 +6,32 @@ class AuthorizationCodeRequest < BaseRequest
       validate :grant,        error: :invalid_grant
       # @see https://tools.ietf.org/html/rfc6749#section-5.2
       validate :redirect_uri, error: :invalid_grant
+      validate :code_verifier, error: :invalid_grant
 
-      attr_accessor :server, :grant, :client, :redirect_uri, :access_token
+      attr_accessor :server, :grant, :client, :redirect_uri, :access_token,
+                    :code_verifier
 
       def initialize(server, grant, client, parameters = {})
         @server = server
         @client = client
         @grant  = grant
         @redirect_uri = parameters[:redirect_uri]
+        @code_verifier = parameters[:code_verifier]
+        @client = client_by_uid(parameters) if no_secret_allowed_for_pkce?
       end
 
       private
 
+      def no_secret_allowed_for_pkce?
+        Doorkeeper.configuration.pkce_without_secret_enabled? &&
+          grant.respond_to?(:code_challenge) &&
+          grant.code_challenge.present? && @client.blank?
+      end
+
+      def client_by_uid(parameters)
+        Doorkeeper::Application.find_by(uid: parameters[:client_id])
```

Comment:
Can you please use `by_uid` method from `Doorkeeper::Application` class ? https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/models/application_mixin.rb#L29